internal/component/statefulset/statefulset.go

```diff
@@ -185,130 +222,113 @@ func (r _resource) getExistingStatefulSet(ctx component.OperatorContext, etcdObj
+				fmt.Sprintf("Error creating or patching StatefulSet with TLS changes for StatefulSet: %v, etcd: %v", client.ObjectKeyFromObject(existingSts), client.ObjectKeyFromObject(etcd)))
+		}
+		return druiderr.New(
+			druiderr.ErrRequeueAfter,
+			component.OperationSync,
+			fmt.Sprintf("Updated TLS config for etcd: %v, requeuing reconcile request", client.ObjectKeyFromObject(etcd)))
+	}
+	peerTLSInSyncForAllMembers, err := utils.IsPeerURLInSyncForAllMembers(ctx, r.client, ctx.Logger, etcd, *existingSts.Spec.Replicas)
 	if err != nil {
 		return druiderr.WrapError(err,
 			ErrSyncStatefulSet,
-			"Sync",
+			component.OperationSync,
 			fmt.Sprintf("Error checking if peer TLS is enabled for statefulset: %v, etcd: %v", client.ObjectKeyFromObject(existingSts), client.ObjectKeyFromObject(etcd)))
 	}
-
-	if !peerTLSEnabledForMembers {
-		if !isStatefulSetPatchedWithPeerTLSVolMount(existingSts) {
-			// This step ensures that only STS is updated with secret volume mounts which gets added to the etcd component due to
-			// enabling of TLS for peer communication. It preserves the current STS replicas.
-			if err = r.createOrPatchWithReplicas(ctx, etcd, *existingSts.Spec.Replicas); err != nil {
-				return druiderr.WrapError(err,
-					ErrSyncStatefulSet,
-					"Sync",
-					fmt.Sprintf("Error creating or patching StatefulSet with TLS enabled for StatefulSet: %v, etcd: %v", client.ObjectKeyFromObject(existingSts), client.ObjectKeyFromObject(etcd)))
-			}
-		} else {
-			ctx.Logger.Info("Secret volume mounts to enable Peer URL TLS have already been mounted. Skipping patching StatefulSet with secret volume mounts.")
-		}
+	if peerTLSInSyncForAllMembers {
+		r.logger.Info("Peer URL TLS configuration is reflected on all currently running members")
+		return nil
+	} else {
 		return druiderr.New(
 			druiderr.ErrRequeueAfter,
-			"Sync",
-			fmt.Sprintf("Peer URL TLS not enabled for #%d members for etcd: %v, requeuing reconcile request", existingSts.Spec.Replicas, client.ObjectKeyFromObject(etcd)))
+			component.OperationSync,
+			fmt.Sprintf("Peer URL TLS not enabled for #%d members for etcd: %v, requeuing reconcile request", *existingSts.Spec.Replicas, client.ObjectKeyFromObject(etcd)))
 	}
-
-	ctx.Logger.Info("Peer URL TLS has been enabled for all currently running members")
-	return nil
 }
 
-func isStatefulSetPatchedWithPeerTLSVolMount(sts *appsv1.StatefulSet) bool {
-	volumes := sts.Spec.Template.Spec.Volumes
-	var peerURLCAEtcdVolPresent, peerURLEtcdServerTLSVolPresent bool
-	for _, vol := range volumes {
-		if vol.Name == common.VolumeNameEtcdPeerCA {
-			peerURLCAEtcdVolPresent = true
-		}
-		if vol.Name == common.VolumeNameEtcdPeerServerTLS {
-			peerURLEtcdServerTLSVolPresent = true
-		}
-	}
-	return peerURLCAEtcdVolPresent && peerURLEtcdServerTLSVolPresent
+func shouldRequeueForMultiNodeEtcdIfPodsNotReady(sts *appsv1.StatefulSet) bool {
+	return sts.Spec.Replicas != nil &&
+		*sts.Spec.Replicas > 1 &&
+		sts.Status.ReadyReplicas > 0 &&
```

Comment:
I think this condition should not be there. If in the cluster, all 3 pods are unready, then ideally `shouldRequeueForMultiNodeEtcdIfPodsNotReady` should return true. But in this case, it ends up returning false, and the reconcile will not be requeued, and the peer TLS gets updated, leading to a hung state for the cluster.

Reply:
In an out-of-band discussion, @unmarshall and I tested and saw that this check is required, to ensure that rolling of a multi-node etcd with enablement of peer TLS works as expected.